src/tacacs/bash_tacplus/bash_tacplus.c

```diff
@@ -350,40 +350,39 @@ int is_local_user(char *user)
     }
 
     struct passwd pwd;
-    struct passwd *pwdresult;
-    char *buf;
-    size_t bufsize = sysconf(_SC_GETPW_R_SIZE_MAX);
-    if (bufsize == -1) {
-        bufsize = DEFAULT_SC_GETPW_R_SIZE_MAX;
-    }
+    struct passwd *pwp;
+    char buf[DEFAULT_GETPWENT_SIZE_MAX];
+    int pwdresult;
+    int result = ERROR_CHECK_LOCAL_USER;
+    setpwent();
+    while (1) {
+        pwdresult = getpwent_r(&pwd, buf, sizeof(buf), &pwp);
+        if (pwdresult) {
+            // no more pw entry
+            break;
+        }
 
-    buf = malloc(bufsize);
-    if (buf == NULL) {
-       output_error("failed to allocate getpwnam_r buffer.\n");
-       return ERROR_CHECK_LOCAL_USER;
-    }
+        if (strcmp(pwp->pw_name, user) != 0) {
+            continue;
+        }
 
-    int s = getpwnam_r(user, &pwd, buf, bufsize, &pwdresult);
-    int result = IS_LOCAL_USER;
-    if (pwdresult == NULL) {
-        if (s == 0)
-            output_error("get user information user failed, user: %s not found\n", user);
+        // compare passwd entry
+        if (strncmp(pwp->pw_gecos, REMOTE_USER_GECOS_PREFIX, strlen(REMOTE_USER_GECOS_PREFIX)) == 0) {
```

Comment:
>strncmp [](http://example.com/codeflow?start=12&length=7)

You need to use buf size, which at least string length + 1. Otherwise you may check string prefix instead of equality.

Reply:
Fixed by change to strcmp.